pkg/provisioner/common.go

```diff
@@ -0,0 +1,240 @@
+	device.Status.ProvisionPhase = targetStatus
+	diskv1.DiskAddedToNode.SetError(device, "", nil)
+	diskv1.DiskAddedToNode.SetStatusBool(device, false)
+	diskv1.DiskAddedToNode.Message(device, message)
+}
+
+func setCondDiskAddedToNodeTrue(device *diskv1.BlockDevice, message string, targetStatus diskv1.BlockDeviceProvisionPhase) {
+	device.Status.ProvisionPhase = targetStatus
+	diskv1.DiskAddedToNode.SetError(device, "", nil)
+	diskv1.DiskAddedToNode.SetStatusBool(device, true)
+	diskv1.DiskAddedToNode.Message(device, message)
+}
+
+func SetCondDeviceFormattingFail(device *diskv1.BlockDevice, err error) {
+	diskv1.DeviceFormatting.SetError(device, "", err)
+	diskv1.DeviceFormatting.SetStatusBool(device, false)
+}
+
+// DiskTags is a cache mechanism for the blockdevices Tags (spec.Tags), it only changed from Harvester side.
+type DiskTags struct {
+	diskTags    map[string][]string
+	lock        *sync.RWMutex
+	initialized bool
+}
+
+func NewLonghornDiskTags() *DiskTags {
+	return &DiskTags{
+		diskTags:    make(map[string][]string),
+		lock:        &sync.RWMutex{},
+		initialized: false,
+	}
+}
+
+func (d *DiskTags) DeleteDiskTags(dev string) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	delete(d.diskTags, dev)
+}
+
+func (d *DiskTags) UpdateDiskTags(dev string, tags []string) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	d.diskTags[dev] = tags
+}
+
+func (d *DiskTags) UpdateInitialized() {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	d.initialized = true
+}
+
+func (d *DiskTags) Initialized() bool {
+	d.lock.RLock()
+	defer d.lock.RUnlock()
+
+	return d.initialized
+}
+
+func (d *DiskTags) GetDiskTags(dev string) []string {
+	d.lock.RLock()
+	defer d.lock.RUnlock()
+
+	return d.diskTags[dev]
+}
+
+func (d *DiskTags) DevExist(dev string) bool {
+	d.lock.RLock()
+	defer d.lock.RUnlock()
+
+	_, found := d.diskTags[dev]
+	return found
+}
+
+// semaphore is a simple semaphore implementation in channel
+type Semaphore struct {
+	ch chan struct{}
+}
+
+// newSemaphore creates a new semaphore with the given capacity.
+func NewSemaphore(n uint) *Semaphore {
+	return &Semaphore{
+		ch: make(chan struct{}, n),
+	}
+}
+
+// acquire a semaphore to prevent concurrent update
+func (s *Semaphore) acquire() bool {
+	logrus.Debugf("Pre-acquire channel stats: %d/%d", len(s.ch), cap(s.ch))
+	select {
+	case s.ch <- struct{}{}:
+		return true
+	default:
+		// full
+		return false
+	}
+}
+
+// release the semaphore
+func (s *Semaphore) release() bool {
+	select {
+	case <-s.ch:
+		return true
+	default:
+		// empty
+		return false
+	}
+}
+
+func valueExists(value string) bool {
+	return value != "" && value != ghwutil.UNKNOWN
+}
+
+func convertMountStr(mountOP NeedMountUpdateOP) string {
+	switch mountOP {
+	case NeedMountUpdateNoOp:
+		return "No-Op"
+	case NeedMountUpdateMount:
+		return "Mount"
+	case NeedMountUpdateUnmount:
+		return "Unmount"
+	}
+	return "Unknown OP"
```

Comment:
I think we could make use of default?

```suggestion
	case NeedMountUpdateUnmount:
		return "Unmount"
	default:
		return "Unknown OP"
	}
```

Reply:
sure, updated!